### PR TITLE
CSPL-1219: Adding app framework boundary tests

### PR DESCRIPTION
**Pull request by jambrosiano**
1 new test added, 1 test updated:

Updated test: The 1st test is updated to perform several apps operation at the same time instead of just updating apps:
This test uses the following apps:
App A: 	microsoft-azure-add-on-for-splunk_300.tgz  		
App B:	palo-alto-networks-add-on-for-splunk_660.tgz  	
App C:	splunk-common-information-model-cim_4181.tgz  	
App D:	splunk-es-content-update_3200.tgz
App E:	add-on-for-ldap_400.tgz

At 1st, it will install A,B,C and D
After install is done, the following operations are performed in a 2nd time:
- Update A and B
- No changes for C
- Remove D from the list of apps
- Install E

Successful runs:
S1:
[2] • [SLOW TEST:1768.474 seconds]
[2] s1appfw test
[2] /Users/jambrosiano/splunk-operator/CSPL-1219/splunk-operator/test/s1/appframework/appframework_test.go:30
[2]   appframework Standalone deployment (S1) with App Framework
[2]   /Users/jambrosiano/splunk-operator/CSPL-1219/splunk-operator/test/s1/appframework/appframework_test.go:64
[2]     smoke, s1, appframework: can deploy a standalone instance with App Framework enabled
[2]     /Users/jambrosiano/splunk-operator/CSPL-1219/splunk-operator/test/s1/appframework/appframework_test.go:65
[2] ------------------------------
[2] {"level":"info","ts":1629941209.110615,"msg":"testenv deleted.\n","testenv":"s1appfw-rh3"}
[2]
[2] JUnit report was created: /Users/jambrosiano/splunk-operator/CSPL-1219/splunk-operator/test/s1/appframework/s1appfw-rh3_junit.xml
[2]
[2] Ran 1 of 1 Specs in 1783.687 seconds
[2] SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
[2] PASS

C3:
[2] • [SLOW TEST:5337.306 seconds]
[2] c3appfw test
[2] /Users/jambrosiano/splunk-operator/CSPL-1219/splunk-operator/test/c3/appframework/appframework_test.go:29
[2]   Single Site Indexer Cluster with SHC (C3) with App Framework
[2]   /Users/jambrosiano/splunk-operator/CSPL-1219/splunk-operator/test/c3/appframework/appframework_test.go:63
[2]     appfwint, c3, appframework: can deploy a C3 SVA with App Framework enabled
[2]     /Users/jambrosiano/splunk-operator/CSPL-1219/splunk-operator/test/c3/appframework/appframework_test.go:64
[2] ------------------------------
[2] {"level":"info","ts":1629939057.600124,"msg":"testenv deleted.\n","testenv":"c3appfw-pde"}
[2] {"level":"info","ts":1629939057.6001651,"msg":"testenv deleted.\n","testenv":"c3appfw-pde"}
[2]
[2] JUnit report was created: /Users/jambrosiano/splunk-operator/CSPL-1219/splunk-operator/test/c3/appframework/c3appfw-pde_junit.xml
[2]
[2] Ran 1 of 1 Specs in 5352.625 seconds
[2] SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
[2] PASS

New test: Installing a bigger amount of apps at the same time
Successful runs:
S1:
[1] • [SLOW TEST:464.035 seconds]
[1] s1appfw test
[1] /Users/jambrosiano/splunk-operator/CSPL-1219/splunk-operator/test/s1/appframework/appframework_test.go:30
[1]   appframework Standalone deployment (S1) with App Framework
[1]   /Users/jambrosiano/splunk-operator/CSPL-1219/splunk-operator/test/s1/appframework/appframework_test.go:339
[1]     smoke, s1, appframework: can deploy a standalone instance with App Framework enabled and install around 350MB of apps at once
[1]     /Users/jambrosiano/splunk-operator/CSPL-1219/splunk-operator/test/s1/appframework/appframework_test.go:340
[1] ------------------------------
[1] {"level":"info","ts":1629422367.995358,"msg":"testenv deleted.\n","testenv":"s1appfw-kex"}
[1]
[1] JUnit report was created: /Users/jambrosiano/splunk-operator/CSPL-1219/splunk-operator/test/s1/appframework/s1appfw-kex_junit.xml
[1]
[1] Ran 1 of 4 Specs in 480.768 seconds
[1] SUCCESS! -- 1 Passed | 0 Failed | 3 Pending | 0 Skipped
[1] PASS

C3:
[1]
[1] • [SLOW TEST:597.546 seconds]
[1] c3appfw test
[1] /Users/jambrosiano/splunk-operator/CSPL-1219/splunk-operator/test/c3/appframework/appframework_test.go:29
[1]   Clustered deployment (C3 - clustered indexer, search head cluster)
[1]   /Users/jambrosiano/splunk-operator/CSPL-1219/splunk-operator/test/c3/appframework/appframework_test.go:685
[1]     smoke, c3, appframework: can deploy a C3 SVA instance with App Framework enabled and install above 200MB of apps at once
[1]     /Users/jambrosiano/splunk-operator/CSPL-1219/splunk-operator/test/c3/appframework/appframework_test.go:686
[1] ------------------------------
[1] {"level":"info","ts":1629421191.313681,"msg":"testenv deleted.\n","testenv":"c3appfw-wzu"}
[1] {"level":"info","ts":1629421191.313709,"msg":"testenv deleted.\n","testenv":"c3appfw-wzu"}
[1]
[1] JUnit report was created: /Users/jambrosiano/splunk-operator/CSPL-1219/splunk-operator/test/c3/appframework/c3appfw-wzu_junit.xml
[1]
[1] Ran 1 of 6 Specs in 613.485 seconds
[1] SUCCESS! -- 1 Passed | 0 Failed | 5 Pending | 0 Skipped
[1] PASS

Note: 
2 apps (Splunk_TA_nix and the splunk_app_stream suite) used for the S1 test are not used in the C3 one as they lead to test failure, probably due to extra setup needed.

### CODE REVIEW

**Comment by sgontla** [comment]
Can we also cover:
1. Re-add the App, that was deleted
2. Upgrade the app:
    2.1. Using the same app package name
    2.2. Using a modified app package name
3. (Not sure if we already have, otherwise)In standalone case
    3.1 Start CR with replicas > 1, and apps
    3.2 Scale up the replicas and make sure that the apps are same across all the Pods.

**Review by tpereirasplunk — APPROVED** [review]

**Comment by jambrosiano** [comment]
> Thanks, @jambrosiano for covering these boundary test cases. In addition to these, can we also cover:
> 
> 1. Re-add the App, that was deleted
> 2. Upgrade the app:
>    2.1. Using the same app package name
>    2.2. Using a modified app package name
> 3. (Not sure if we already have, otherwise)In standalone case
>    3.1 Start CR with replicas > 1, and apps
>    3.2 Scale up the replicas, and also have the add/modify/delete combination, and make sure that the apps are the same across all the Pods.

Thanks ,@sgontla for bringing these test ideas. 
Here's the status for each of them:
1:  this test could be added in a new PR.
2.1 this is about having 2 files with exact same name, but different content/hash, like an updated .conf file for ex. This could be added also in a new PR.
2.2 This is handled by the current update tests we have.
3.1: This could be done by a PR verifying CSPL-1273, I'll create a ticket for that.
3.2: This is covered by the updated test in this PR, as it goes this way: install V1-> Upgrade to V2 -> Scale -> Downgrade to V1  -> do the add/modify/delete combination

File: test/s1/appframework/appframework_test.go [hunk 1]
```diff
@@ -287,4 +328,70 @@ var _ = Describe("s1appfw test", func() {
 			testenv.VerifyAppInstalled(deployment, testenvInstance, testenvInstance.GetName(), standalonePod, esApp, false, "enabled", false, false)
 		})
 	})
+
+	Context("appframework Standalone deployment (S1) with App Framework", func() {
+		It("smoke, s1, appframework: can deploy a standalone instance with App Framework enabled and install around 350MB of apps at once", func() {
```
Comment (pdhanoyasplunk) [inline test/s1/appframework/appframework_test.go:333]: Nit: This is redundant as ES app is already 2GB +?

File: test/c3/appframework/appframework_test.go [hunk 2]
```diff
@@ -278,18 +278,78 @@ var _ = Describe("c3appfw test", func() {
...
-			testenvInstance.Log.Info("Verify Apps are NOT copied to /etc/apps on CM and Deployer for app", " verison", appVersion)
+			testenvInstance.Log.Info("Verify Apps are NOT copied to /etc/apps on CM and Deployer for app", " version", appVersion)
 			testenv.VerifyAppsCopied(deployment, testenvInstance, testenvInstance.GetName(), masterPodNames, appListV1, false, false)
 
-			//Verify Apps are updated
+			// Verify Apps are updated
 			testenvInstance.Log.Info("Verify Apps are installed on the pods by running Splunk CLI commands for app", " version", appVersion)
 			testenv.VerifyAppInstalled(deployment, testenvInstance, testenvInstance.GetName(), allPodNames, appListV1, true, "enabled", false, true)
 
+			// Delete apps on S3 for new Apps
```
Comment (pdhanoyasplunk) [inline test/c3/appframework/appframework_test.go:293]: Nit: The run time of this test is already 30 mins + .. this would take it into 45 min mark.

**Review by pdhanoyasplunk — APPROVED** [review]

File: test/s1/appframework/appframework_test.go (hunk 1 above)
Comment (jambrosiano) [reply test/s1/appframework/appframework_test.go:333]: ES app is a special case, and installed alone. Here we install several apps together, more than just the basic apps

File: test/c3/appframework/appframework_test.go (hunk 2 above)
Comment (jambrosiano) [reply test/c3/appframework/appframework_test.go:293]: This test becomes indeed very long, I would suggest this to be the very last addition to it.

**Review by sgontla — APPROVED** [review]

File: test/c3/appframework/appframework_test.go (hunk 2 above)
Comment (pdhanoyasplunk) [reply test/c3/appframework/appframework_test.go:293]: Ideally this should not be added to the test however we have separate task to break these test down.

Comment (smohan-splunk) [reply test/c3/appframework/appframework_test.go:293]: Thankfully its integration and not smoke!!

**Review by smohan-splunk — APPROVED** [review]

**Pull request merged**